packages/react/ds-core/src/ui/Tooltip/Tooltip.tsx

```diff
@@ -0,0 +1,264 @@
+    let yOffset = 0;
+
+    switch (position) {
+      case "topCenter":
+      case "topRight":
+      case "topLeft":
+        yOffset = -(messageRect.height + ARROW_SIZE);
+        break;
+      case "btmCenter":
+      case "btmRight":
+      case "btmLeft":
+        yOffset = targetRect.height + ARROW_SIZE;
+        break;
+      case "left":
+      case "right":
+        yOffset = (targetRect.height - messageRect.height) / 2;
+        break;
+    }
+
+    switch (position) {
+      case "topCenter":
+      case "btmCenter":
+        xOffset = (targetRect.width - messageRect.width) / 2;
+        break;
+      case "left":
+        xOffset = -(messageRect.width + ARROW_SIZE);
+        break;
+      case "right":
+        xOffset = targetRect.width + ARROW_SIZE;
+        break;
+      case "topRight":
+      case "btmRight":
+        xOffset = targetRect.width - messageRect.width + ARROW_SIZE;
+        break;
+      case "topLeft":
+      case "btmLeft":
+        xOffset = -ARROW_SIZE;
+        break;
+    }
+
+    let adjustedXOffset = xOffset;
+    let adjustedYOffset = yOffset;
+
+    if (autoAdjust) {
+      const viewportWidth = window.innerWidth;
+      const viewportHeight = window.innerHeight;
+
+      const absoluteX = targetRect.left + xOffset;
+      const absoluteY = targetRect.top + yOffset;
+
+      if (absoluteX < 0) adjustedXOffset -= absoluteX;
+      else if (absoluteX + messageRect.width > viewportWidth)
+        adjustedXOffset -= absoluteX + messageRect.width - viewportWidth;
+
+      if (absoluteY < 0) adjustedYOffset -= absoluteY;
+      else if (absoluteY + messageRect.height > viewportHeight)
+        adjustedYOffset -= absoluteY + messageRect.height - viewportHeight;
+    }
+    setMessagePosition({
+      top: `${adjustedYOffset}px`,
+      left: `${adjustedXOffset}px`,
+    });
+  }, [position, autoAdjust]);
+
+  useEffect(() => {
+    calculateMessagePosition();
+  }, [calculateMessagePosition]);
+
+  useEffect(() => {
+    const handleResize = () => {
+      if (isVisible) calculateMessagePosition();
+    };
+
+    const handleScroll = () => {
+      if (isVisible) calculateMessagePosition();
+    };
+
+    window.addEventListener("resize", handleResize);
+    window.addEventListener("scroll", handleScroll, { passive: true });
+
+    return () => {
+      window.removeEventListener("resize", handleResize);
+      window.removeEventListener("scroll", handleScroll);
+    };
+  }, [isVisible, calculateMessagePosition]);
+
+  useEffect(() => {
+    if (isVisible && messageRef.current && isFocused) {
+      const firstFocusableElement = messageRef.current.querySelector(
+        '[tabIndex="0"], a, button, input, textarea, select',
+      ) as HTMLElement;
+      if (firstFocusableElement) firstFocusableElement.focus();
+    }
+  }, [isVisible, isFocused]);
+
+  const replaceTimer = useCallback(
+    (newTimer?: ReturnType<typeof setTimeout>) => {
+      if (visibilityTimeout) clearTimeout(visibilityTimeout);
```

Comment:
you could use a `useRef` for a timer which would be better in terms of performance if you don't need the reactivity part in `visibilityTimeout`.

Reply:
Good call, implemented!